connections/inlined-muxer-negotiation.md

```diff
@@ -0,0 +1,128 @@
+# Stream multiplexer negotiation in security handshake <!-- omit in toc -->
+
+
+| Lifecycle Stage | Maturity      | Status | Latest Revision |
+|-----------------|---------------|--------|-----------------|
+| 1A              | Working Draft | Active | r1, 2022-12-07  |
+
+Authors: [@julian88110], [@marten-seemann]
+
+Interest Group: [@marcopolo], [@mxinden]
+
+[@marten-seemann]: https://github.com/marten-seemann
+[@marcopolo]: https://github.com/marcopolo
+[@mxinden]: https://github.com/mxinden
+[@julian88110]: https://github.com/julian88110
+
+See the [lifecycle document][lifecycle-spec] for context about maturity level
+and spec status.
+
+[lifecycle-spec]: https://github.com/libp2p/specs/blob/master/00-framework-01-spec-lifecycle.md
+
+## Table of Contents <!-- omit in toc -->
+
+- [Overview](#overview)
+- [Design](#design)
+  - [Multiplexer Negotiation over TLS](#multiplexer-negotiation-over-tls)
+  - [Multiplexer Negotiation over Noise](#multiplexer-negotiation-over-noise)
+- [Privacy](#privacy)
+- [Alternative options considered](#alternative-options-considered)
+
+## Overview
+
+Transports that don't support native stream multiplexing (e.g. TCP, WebSocket) negotiate
+a stream multiplexer after completion of the cryptographic handshake, as described in [connections]. 
+Negotiating the stream multiplexer takes one network roundtrip.
+This document defines a backwards-compatible optimization, which allows running the
+multiplexer negotiation during the cryptographic handshake, thereby reducing the latency of
+connection establishment by one roundtrip.
+
+
+## Design
+
+### Multiplexer Negotiation over TLS
+
+When using TLS, the [ALPN] extension is used to negotiate the multiplexer.
+
+The ALPN TLS extension allows the client to send a list of supported application
+protocols as part of the TLS ClientHello message.  The server chooses
+a protocol and sends the selected protocol as part of the TLS
+ServerHello message.
+
+For the purpose of multiplexer negotiation, the protocol IDs of the stream 
+multiplexers are sent, followed by the `"libp2p"` protocol code. The multiplexer list is ordered by
+the client's preference, with the most preferred multiplexer at the beginning.
+The server SHOULD respect the client's preference and pick the first protocol
+from the list that it supports.
+
+Example for a node supporting both yamux and mplex, with a preference for yamux:
+```json
+[ "/yamux/1.0.0", "/mplex/6.7.0", "libp2p" ]
+```
+
+The `"libp2p"` protocol code MUST always be the last item in the multiplexer list.
+According to [TLS], nodes that don't implement the optimization described in
+this document use `"libp2p"` for their ALPN. If `"libp2p"` is the result of the
+ALPN process, nodes MUST use multistream negotiation of the stream multiplexer
+as described in [connections].
```

Comment:
This is to enable backwards-compatibility with legacy implementations. With these nodes, the result of the negotiation process will be "libp2p", so we need to do multistream.
Now we could say that we fail the handshake if there's anything other than "libp2p" in the list, but this would make it impossible to deploy a listener that speaks HTTP and libp2p without this optimization.